File: c/all_permutations_of_string.c
#include <stdio.h>

void swap(char *x, char *y) {
    char temp = *x;
    *x = *y;
    *y = temp;
}

void permute(char *str, int start, int end) {
    int j = start;
    if (start == end) {
        printf("\t%s\n", str);
    } else {
        for (j = start; j <= end; j++) {
            swap((str + start), (str + j));
            permute(str, start + 1, end);
            swap((str + start), (str + j));
        }
    }
}

void sample1() {
    char a[] = "abc";
    printf("\n--- Permutations for %s: \n", a);
    permute(a, 0, 2);
}

void sample2() {
    char a[] = "abcd";
    printf("\n--- Permutations for %s: \n", a);
    permute(a, 0, 3);
}

int main() {
    sample1();
    sample2();
    return 0;
}

File: c_sharp/all_permutations_of_string.cs
using System;

namespace blogposts
{
    public class Permuter
    {
        private int start;
        private int end;
        private string current;
        public Permuter(string str, int startIndex, int endIndex)
        {
            current = str;
            start = startIndex;
            end = endIndex;
        }

        private string Swap(string str, int i, int j)
        {
            char[] mutableStr = str.ToCharArray();
            char temp = mutableStr[i];
            mutableStr[i] = mutableStr[j];
            mutableStr[j] = temp;
            return new string(mutableStr);
        }

        public void Permute()
        {
            if (start == end)
                Console.WriteLine(current);
            else
            {
                for (int j = start; j <= end; j++)
                {
                    string tmp = Swap(current, start, j);
                    (new Permuter(tmp, start + 1, end)).Permute();
                }
            }
        }
    }

    class Runner{
        static void Main(){
            Permuter p = new Permuter("abc", 0, 2);
            p.Permute();
        }
    }
}